extensions/amp-story-shopping/0.1/amp-story-shopping-attachment.js

```diff
@@ -14,6 +25,70 @@ export class AmpStoryShoppingAttachment extends AmpStoryPageAttachment {
 
     /** @private {?Element} */
     this.container_ = null;
+
+    /** @private @const {?../../amp-story/1.0/amp-story-store-service.AmpStoryStoreService} */
+    this.storeService_ = null;
+
+    /** @private {?../../../src/service/localization.LocalizationService} */
+    this.localizationService_ = null;
+  }
+
+  /**
+   * @private
+   * @return {?Element} the parent amp-story-page
+   */
+  getPage_() {
+    return closest(
+      dev().assertElement(this.element),
+      (el) => el.tagName.toLowerCase() === 'amp-story-page'
+    );
+  }
+
+  /**
+   * Updates the CTA based on the shopping data.
+   * @private
+   */
+  updateCtaText_() {
+    const pageIdElem = this.getPage_();
+
+    const shoppingTags = Array.from(
+      pageIdElem.getElementsByTagName('amp-story-shopping-tag')
+    ).filter(
+      (tag) =>
+        this.storeService_.get(StateProperty.SHOPPING_DATA)[
+          tag.getAttribute('data-tag-id')
+        ]
+    );
+    const numShoppingTags = shoppingTags.length;
+
+    const ctaButton = pageIdElem.querySelector(
+      '.i-amphtml-story-page-open-attachment-host'
+    )?.shadowRoot.children[1];
+
+    if (numShoppingTags === 0 || !ctaButton) {
+      // Failsafe in case the CTA button fails to load, will call the function again until CTA button loads.
+      setTimeout(() => this.updateCtaText_(), 100);
```

Comment:
We must always keep performance in mind when working on our codebase. We cannot have timeouts running.
We need to find a way to only update the button once.
There must be another way. 

The cta button sets it's text using the `data-cta-text` attribute. Maybe we can try setting that attribute if the button is not built.

codepointer to where this happens in when building the CTA:
https://github.com/ampproject/amphtml/blob/main/extensions/amp-story/1.0/amp-story-open-page-attachment.js#L156

Reply:
After a ton of testing and debugging, it turns out that we don't even need to do this - the bug was being caused by the listener not calling the page change sometimes when the page is not loaded.

I added a call to this.updateCtaText_(); in the buildCallback and it now seems to work.

I recall the following comment from `amp-story-auto-ads` (where I got the page change listener pattern from)

"Respond to page navigation event. This method is not called for the first
 page that is shown on load."
   
I'm going to write some unit tests to test this out some more, but I think we've reached a point where we have consistently predictable behavior.